src/Workspaces/Core/Portable/Utilities/AsyncLazy`1.cs

```diff
@@ -372,18 +372,18 @@ private void StartAsynchronousComputation(AsynchronousComputationToStart computa
                 // a state that was complete.
                 try
                 {
-                    // We avoid creating a full closure just to pass the token along
-                    // Also, use TaskContinuationOptions.ExecuteSynchronously so that we inline 
-                    // the continuation if asynchronousComputeFunction completes synchronously
                     var task = computationToStart.AsynchronousComputeFunction(cancellationToken);
 
-                    task.ContinueWith(
-                        (t, s) => CompleteWithTask(t, ((CancellationTokenSource)s).Token),
-                        computationToStart.CancellationTokenSource,
-                        cancellationToken,
-                        TaskContinuationOptions.ExecuteSynchronously,
-                        TaskScheduler.Default);
-
+                    // As an optimization, if the task is already completed, mark the 
+                    // request as being completed as well.
+                    //
+                    // Note: we want to do this before we do the .ContinueWith below. That way, 
+                    // when the async call to CompleteWithTask runs, it sees that we've already
+                    // completed and can bail immediately.  If we were to do this after we 
+                    // kicked off the async work, then we'd have the chance that both would
+                    // run concurrently and we'd have a higher change of hitting the race condition
+                    // of calling TaskCompletionSource.SetResult simultaneously (and thus having 
```

Comment:
Note: it's not TaskCompletionSource but AsyncMethodBuilder. If it was TaskCompletionSource then we wouldn't have this problem. Please amend the commit since once we move to TaskCompletionSource in 4.6 we'll probably want to note this change and undo it in it's entirety since I guess technically you'd want the async work kicked off first if you had the choice so you could complete other tasks a microsecond earlier. But any such benefit there is totally overshadowed by the annoyingness of the exception today.